### PR TITLE
node - add generics to the strict equal assertions

### PR DESCRIPTION
**Pull request by DetachHead**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [equivalent assertions in chai](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/468ef892e6b8271b4ebe67ffb9f484f9c6fa6fc0/types/chai/index.d.ts#L486)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DetachHead** [comment]
i fixed the failures caused by my changes. the tests are still failing due to https://github.com/microsoft/TypeScript/commit/bce412cb15d7e1229ec02169276e481b60079b7a

see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50262#issuecomment-750676745

**Comment by calebsander** [comment]
@DetachHead It seems like the latest nightly version of TypeScript has fixed these type errors in the declaration files.

**Comment by SimonSchick** [comment]
What problem does this solve?

**Comment by DetachHead** [comment]
@SimonSchick 
currently, there is no compiler error on invalid deep equality assertions. for example:

```ts
const foo: Foo = {
    bar: 1
}

assert.deepStrictEqual(foo, { bar: "" }) //this can never pass, the types don't match so it should be a compiler error

File: types/node/node-tests.ts [hunk 1]
```diff
@@ -52,12 +52,12 @@ assert['fail'](true, true, 'works like a charm');
 
 {
     const a = 'test' as any;
-    assert.strictEqual(a, 'test');
+    assert.strictEqual<string>(a, 'test');
```
Comment (Flarna) [inline types/node/node-tests.ts:55]: Seems this change causes that asserts like this require users to specify the type.
I don't think this is helpful.

Comment (DetachHead) [reply types/node/node-tests.ts:55]: i don't see why the user should be allowed to specify values with non-matching types without having to explicitly specify that they're intending to do so (either via type casting or specifying generics). it seems like the cost of the amount of runtime errors that could easily be picked up at compiletime (like my [example above](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50274#issuecomment-753414752)) outweighs the slight convenience for edge cases where you'd actually want non-matching types.

that said, your example could be fixed by doing something like this, but i question whether doing so is a good idea
```ts
function strictEqual<T extends U, U>(actual: U, expected: T, message?: string | Error): asserts actual is T
```

Comment (Flarna) [reply types/node/node-tests.ts:55]: It's not about wanting non matching types. `any` and `string` match similar as `string | number` matches `string`.
But not the assert is no longer asserting these types correct.

Using `function strictEqual<T extends U, U>(actual: U, expected: T, message?: string | Error): asserts actual is T` sounds better as it solves at least the cases with `any` and union types which are quite common in my opinion.

Comment (DetachHead) [reply types/node/node-tests.ts]: done

File: types/libpq/tslint.json [hunk 2]
```diff
@@ -2,6 +2,7 @@
     "extends": "dtslint/dt.json",
     "rules": {
         "ban-types": false,
-        "no-void-expression": false
+        "no-void-expression": false,
+        "no-outside-dependencies": false
```
Comment (Flarna) [inline types/libpq/tslint.json:6]: Why is this needed?

Comment (DetachHead) [reply types/libpq/tslint.json:6]: https://github.com/microsoft/dtslint/issues/315

**Comment by Flarna** [comment]
I tried this change in my project and got 110 compile errors which are all invalid and would require an unneeded cast or other code modifications.

Quite some of them are related to `undefined` handling like this
```
interface Foo {
  [x: string]: string
}
const f: Foo = {}
assert.strictEqual(f.bar, undefined); // Argument of type 'undefined' is not assignable to parameter of type 'string'.
```

But there are also others where one assert has side effects to the next like this simplified sample:
```
class Bar {
    get f(): string | undefined { return this._f }
    update(): void { this._f = "name" }
    private _f?: string
}
const b = new Bar();
assert.strictEqual(b.f, undefined);
b.update();
assert.strictEqual(b.f, "name"); // Argument of type '"name"' is not assignable to parameter of type 'undefined'.
```

And there are more where current code works fine but typescript would complain.

I think it's not the best choice to add this considering that `@types/node` is used in millions of projects which would be likely broken because of this.

**Comment by DetachHead** [comment]
> I tried this change in my project and got 110 compile errors which are all invalid and would require an unneeded cast or other code modifications.
> 
> Quite some of them are related to `undefined` handling like this
> 
> ```
> interface Foo {
>   [x: string]: string
> }
> const f: Foo = {}
> assert.strictEqual(f.bar, undefined); // Argument of type 'undefined' is not assignable to parameter of type 'string'.
> ```
good point, though using `noUncheckedIndexedAccess` would fix this i suppose that's not a reasonable solution
> But there are also others where one assert has side effects to the next like this simplified sample:
> 
> ```
> class Bar {
>     get f(): string | undefined { return this._f }
>     update(): void { this._f = "name" }
>     private _f?: string
> }
> const b = new Bar();
> assert.strictEqual(b.f, undefined);
> b.update();
> assert.strictEqual(b.f, "name"); // Argument of type '"name"' is not assignable to parameter of type 'undefined'.
> ```
can be solved with
```ts
assert.strictEqual<string | undefined>(b.f, "name");
```
> And there are more where current code works fine but typescript would complain.
> 
> I think it's not the best choice to add this considering that `@types/node` is used in millions of projects which would be likely broken because of this.

fair enough. my reasoning is that although this would be a breaking change i feel it's outweighed by the benefits. so many incorrect assertions i've written could have easily been picked up by the compiler and i'm sure such scenarios are far more common.

**Comment by andrewbranch** [comment]
Thanks @Flarna for testing against a real project, that’s always super helpful in cases like these. I agree this seems risky, and would encourage those who want this kind of behavior to type their own utility function wrapper or add this as an overload via module augmentation. I think given the amount of disruption this is proven to have, there’s no viable path forward. Thanks nonetheless, @DetachHead 👍

**Comment by DetachHead** [comment]
> would encourage those who want this kind of behavior to type their own utility function wrapper or add this as an overload via module augmentation

done - https://www.npmjs.com/package/typed-nodejs-assert
